diff --git a/src/compiler/nir/nir_opt_algebraic.py b/src/compiler/nir/nir_opt_algebraic.py
--- a/src/compiler/nir/nir_opt_algebraic.py
+++ b/src/compiler/nir/nir_opt_algebraic.py
@@ -444,10 +444,21 @@ for sz in (16, 32, 64):
 
 # Shorthand for the expansion of just the dot product part of the [iu]dp4a
 # instructions.
-sdot_4x8_a_b = ('iadd', ('iadd', ('imul24_relaxed', ('extract_i8', a, 0), ('extract_i8', b, 0)),
-                                 ('imul24_relaxed', ('extract_i8', a, 1), ('extract_i8', b, 1))),
-                        ('iadd', ('imul24_relaxed', ('extract_i8', a, 2), ('extract_i8', b, 2)),
-                                 ('imul24_relaxed', ('extract_i8', a, 3), ('extract_i8', b, 3))))
+sdot_4x8_a_b_balanced = ('iadd',
+                    ('iadd',
+                     ('imul24_relaxed', ('extract_i8', a, 0), ('extract_i8', b, 0)),
+                     ('imul24_relaxed', ('extract_i8', a, 1), ('extract_i8', b, 1))),
+                    ('iadd',
+                     ('imul24_relaxed', ('extract_i8', a, 2), ('extract_i8', b, 2)),
+                     ('imul24_relaxed', ('extract_i8', a, 3), ('extract_i8', b, 3))))
+
+sdot_4x8_a_b_reassoc = ('iadd',
+                    ('imul24_relaxed', ('extract_i8', a, 3), ('extract_i8', b, 3)),
+                    ('iadd',
+                     ('imul24_relaxed', ('extract_i8', a, 2), ('extract_i8', b, 2)),
+                     ('iadd',
+                      ('imul24_relaxed', ('extract_i8', a, 1), ('extract_i8', b, 1)),
+                      ('imul24_relaxed', ('extract_i8', a, 0), ('extract_i8', b, 0)))))
 udot_4x8_a_b = ('iadd', ('iadd', ('umul24_relaxed', ('extract_u8', a, 0), ('extract_u8', b, 0)),
                                  ('umul24_relaxed', ('extract_u8', a, 1), ('extract_u8', b, 1))),
                         ('iadd', ('umul24_relaxed', ('extract_u8', a, 2), ('extract_u8', b, 2)),
@@ -462,7 +473,8 @@ udot_2x16_a_b = ('iadd', ('umul24_relaxed', ('extract_u16', a, 0), ('extract_u16
                          ('umul24_relaxed', ('extract_u16', a, 1), ('extract_u16', b, 1)))
 
 optimizations.extend([
-   (('sdot_4x8_iadd', a, b, c), ('iadd', sdot_4x8_a_b, c), '!options->has_sdot_4x8'),
+   (('sdot_4x8_iadd', a, b, '#c'), ('iadd', sdot_4x8_a_b_balanced, c), '!options->has_sdot_4x8'),
+   (('sdot_4x8_iadd', a, b, c), ('iadd', sdot_4x8_a_b_reassoc, c), '!options->has_sdot_4x8'),
    (('udot_4x8_uadd', a, b, c), ('iadd', udot_4x8_a_b, c), '!options->has_udot_4x8'),
    (('sudot_4x8_iadd', a, b, c), ('iadd', sudot_4x8_a_b, c), '!options->has_sudot_4x8'),
    (('sdot_2x16_iadd', a, b, c), ('iadd', sdot_2x16_a_b, c), '!options->has_dot_2x16'),
